### PR TITLE
perf(pixel): optimize plasma batch hot paths

### LINKED ISSUE #47: [Performance] Benchmark the existing PlasmaBrightnessContrast batch path
## Current status

The batch implementation is already present.

`PlasmaBrightnessContrast.apply_to_images` uses `@batch_transform("spatial")` to reshape `(N, H, W, C)` into one spatial operation and calls `apply` once. It does not iterate over images in Python. The plasma pattern is generated once and shared across the batch.

The generic `images=` coverage currently passes for this transform.

## Remaining work

Before closing this issue:

- add or identify focused parity coverage for the native batch path;
- benchmark the existing native path against the inherited per-image fallback;
- confirm that reshape and restore allocations do not erase the benefit for small batches;
- document any workload where the fallback remains faster and add a guarded selection if needed.

## Benchmark requirements

Compare direct `apply_to_images` and `Compose(images=...)` under one CPU thread.

Benchmark 256×256, 512×512, and 1024×1024 inputs; 1, 3, and 5 channels; `uint8` and `float32`; and batch sizes 2, 4, 8, and 16. Report every result. A regression greater than 5% requires justification or rework.

If the current path passes parity and performance verification, close this issue as completed.

### PR DESCRIPTION
**Pull request by ayasyrev**
Closes #47.

## Summary

`PlasmaBrightnessContrast.apply_to_images` previously folded `(N, H, W, C)` into one image. That changed the reduction domain for contrast: the functional computed one batch-wide mean instead of one mean per image.

This PR replaces that route with an exact hybrid dispatcher while keeping the public API, sampled parameters, and single-image implementation unchanged:

- nonzero singleton batches use the existing single-image kernel;
- zero-effect batches return an owned, writable, C-contiguous semantic copy, with float32 clipping performed in one pass;
- float32 batches of size 2–4 use a dedicated 4-D kernel with one mean per image;
- uint8 and larger float32 batches use a preallocated per-image loop with shared plasma maps;
- high-resolution, five-or-more-channel uint8 brightness-only batches of size 2–4 use the inherited per-image route where it is faster.

## Correctness

The oracle applies the unchanged single-image functional independently to every item with identical sampled parameters and stacks the results. Focused coverage includes uint8/float32, 1/3/5 channels, implicit grayscale, `images`, `volume`, empty and singleton batches, zero/brightness-only/contrast-only/combined factors, read-only and non-contiguous inputs, input ownership, writable C-contiguous output, and route selection.

The production implementation was exact in all 144 issue-matrix comparisons against the inherited per-image fallback. The focused functional test currently adds 48 exact parity and non-mutation cases.

## Performance method

- Baseline: faithful inherited per-image fallback at AlbumentationsX `91ccf9e` (`2.4.3`).
- Candidate: production hybrid dispatcher, in the same process and environment.
- Alternating medians, five samples per implementation, automatic loop count targeting 80 ms (maximum 16 loops).
- One OpenCV thread and controlled BLAS/OpenMP/Torch thread counts.
- Matrix: 256/512/1024 square inputs; 1/3/5 channels; uint8/float32; batch sizes 2/4/8/16; direct `apply_to_images` and public `Compose(images=...)`.
- Host: AMD Ryzen 7 5800X, 8 cores/16 threads; Python 3.12.13, NumPy 2.5.1, OpenCV 5.0.0, Albucore 0.2.14.

| Route | Cells | Geometric-mean speedup | Minimum speedup | Regressions over 5% |
| --- | ---: | ---: | ---: | ---: |
| Direct `apply_to_images` | 72 | 1.340x | 0.995x | 0/72 |
| `Compose(images=...)` | 72 | 1.313x | 1.010x | 0/72 |

The pure dedicated candidate was rejected: it reached only 0.852x direct / 0.865x Compose geometric-mean speedup, regressed in 36/72 cells on each route, and increased largest-case peak RSS by 119.8% for uint8 and 21.6% for float32. This is why the whole-batch kernel is bounded to float32 batches of at most four items.

Largest-case production peak RSS (`1024x1024x5`, N=16) stayed within the 5% budget:

| Route | Dtype | Fallback | Production | Change |
| --- | --- | ---: | ---: | ---: |
| Direct | uint8 | 652.1 MiB | 667.1 MiB | +2.30% |
| Direct | float32 | 1106.5 MiB | 1107.4 MiB | +0.08% |
| Compose | uint8 | 659.5 MiB | 674.6 MiB | +2.28% |
| Compose | float32 | 1114.2 MiB | 1114.5 MiB | +0.02% |

<details>
<summary>Complete 144-result issue matrix (milliseconds, fallback / production)</summary>

| Size | C | Dtype | N | Direct F / P | Compose F / P |
| --- | ---: | --- | ---: | ---: | ---: |
| 256 | 1 | uint8 | 2 | 1.088 / 0.996 | 2.523 / 1.579 |
| 256 | 1 | uint8 | 4 | 2.839 / 1.255 | 3.501 / 1.915 |
| 256 | 1 | uint8 | 8 | 6.976 / 1.765 | 6.364 / 2.351 |
| 256 | 1 | uint8 | 16 | 3.715 / 2.974 | 4.473 / 3.639 |
| 256 | 1 | float32 | 2 | 1.107 / 1.005 | 1.669 / 1.648 |
| 256 | 1 | float32 | 4 | 1.559 / 1.275 | 2.310 / 1.927 |
| 256 | 1 | float32 | 8 | 1.075 / 0.798 | 3.173 / 1.499 |
| 256 | 1 | float32 | 16 | 2.168 / 1.575 | 2.902 / 2.206 |
| 256 | 3 | uint8 | 2 | 0.944 / 0.845 | 1.627 / 1.503 |
| 256 | 3 | uint8 | 4 | 1.865 / 1.466 | 5.477 / 5.055 |
| 256 | 3 | uint8 | 8 | 6.827 / 5.776 | 7.335 / 6.270 |
| 256 | 3 | uint8 | 16 | 11.173 / 8.785 | 11.725 / 9.314 |
| 256 | 3 | float32 | 2 | 3.738 / 0.865 | 4.269 / 1.501 |
| 256 | 3 | float32 | 4 | 5.263 / 4.757 | 5.798 / 5.437 |
| 256 | 3 | float32 | 8 | 6.642 / 5.518 | 7.307 / 6.114 |
| 256 | 3 | float32 | 16 | 6.624 / 5.147 | 7.606 / 5.778 |
| 256 | 5 | uint8 | 2 | 1.416 / 1.288 | 2.100 / 1.943 |
| 256 | 5 | uint8 | 4 | 2.898 / 2.270 | 3.566 / 2.958 |
| 256 | 5 | uint8 | 8 | 5.750 / 4.227 | 6.507 / 4.943 |
| 256 | 5 | uint8 | 16 | 11.596 / 8.194 | 12.525 / 8.963 |
| 256 | 5 | float32 | 2 | 1.240 / 1.080 | 1.927 / 1.765 |
| 256 | 5 | float32 | 4 | 2.586 / 1.929 | 3.371 / 2.708 |
| 256 | 5 | float32 | 8 | 5.427 / 3.872 | 6.157 / 4.576 |
| 256 | 5 | float32 | 16 | 10.914 / 8.714 | 11.914 / 8.398 |
| 512 | 1 | uint8 | 2 | 1.066 / 0.887 | 1.802 / 1.624 |
| 512 | 1 | uint8 | 4 | 2.145 / 1.723 | 2.914 / 2.458 |
| 512 | 1 | uint8 | 8 | 4.269 / 3.312 | 5.121 / 4.054 |
| 512 | 1 | uint8 | 16 | 8.494 / 6.547 | 9.517 / 7.346 |
| 512 | 1 | float32 | 2 | 0.906 / 0.711 | 1.652 / 1.464 |
| 512 | 1 | float32 | 4 | 1.833 / 1.346 | 2.628 / 2.131 |
| 512 | 1 | float32 | 8 | 3.846 / 2.820 | 4.636 / 3.581 |
| 512 | 1 | float32 | 16 | 7.832 / 5.853 | 8.704 / 6.569 |
| 512 | 3 | uint8 | 2 | 5.274 / 4.568 | 6.377 / 5.555 |
| 512 | 3 | uint8 | 4 | 10.996 / 8.078 | 11.718 / 9.215 |
| 512 | 3 | uint8 | 8 | 21.778 / 15.026 | 23.245 / 16.079 |
| 512 | 3 | uint8 | 16 | 42.915 / 28.311 | 57.464 / 42.066 |
| 512 | 3 | float32 | 2 | 4.588 / 3.710 | 5.741 / 4.921 |
| 512 | 3 | float32 | 4 | 9.747 / 7.779 | 27.712 / 19.181 |
| 512 | 3 | float32 | 8 | 30.947 / 25.081 | 34.229 / 28.568 |
| 512 | 3 | float32 | 16 | 43.575 / 32.847 | 41.914 / 31.255 |
| 512 | 5 | uint8 | 2 | 23.071 / 11.756 | 26.255 / 23.617 |
| 512 | 5 | uint8 | 4 | 32.062 / 26.142 | 32.745 / 26.616 |
| 512 | 5 | uint8 | 8 | 55.088 / 37.999 | 54.137 / 39.412 |
| 512 | 5 | uint8 | 16 | 101.098 / 65.678 | 103.605 / 67.402 |
| 512 | 5 | float32 | 2 | 16.795 / 15.169 | 18.419 / 16.890 |
| 512 | 5 | float32 | 4 | 27.771 / 21.086 | 29.053 / 22.710 |
| 512 | 5 | float32 | 8 | 39.190 / 34.111 | 40.150 / 34.945 |
| 512 | 5 | float32 | 16 | 77.320 / 55.186 | 79.587 / 54.779 |
| 1024 | 1 | uint8 | 2 | 8.352 / 6.436 | 9.534 / 7.449 |
| 1024 | 1 | uint8 | 4 | 16.426 / 11.097 | 17.912 / 12.361 |
| 1024 | 1 | uint8 | 8 | 33.619 / 21.711 | 44.085 / 23.865 |
| 1024 | 1 | uint8 | 16 | 75.027 / 51.649 | 77.345 / 52.189 |
| 1024 | 1 | float32 | 2 | 7.096 / 5.332 | 16.400 / 7.219 |
| 1024 | 1 | float32 | 4 | 23.367 / 15.332 | 25.321 / 18.490 |
| 1024 | 1 | float32 | 8 | 29.237 / 25.273 | 31.970 / 27.141 |
| 1024 | 1 | float32 | 16 | 58.463 / 38.576 | 59.181 / 41.685 |
| 1024 | 3 | uint8 | 2 | 54.659 / 43.968 | 55.901 / 45.683 |
| 1024 | 3 | uint8 | 4 | 108.090 / 75.663 | 109.100 / 78.670 |
| 1024 | 3 | uint8 | 8 | 215.774 / 139.812 | 216.532 / 141.954 |
| 1024 | 3 | uint8 | 16 | 434.304 / 272.320 | 436.428 / 275.547 |
| 1024 | 3 | float32 | 2 | 39.457 / 35.896 | 40.635 / 38.029 |
| 1024 | 3 | float32 | 4 | 81.476 / 64.643 | 85.496 / 66.536 |
| 1024 | 3 | float32 | 8 | 139.375 / 120.304 | 139.211 / 121.428 |
| 1024 | 3 | float32 | 16 | 272.580 / 229.165 | 273.505 / 228.361 |
| 1024 | 5 | uint8 | 2 | 83.850 / 84.239 | 85.586 / 84.754 |
| 1024 | 5 | uint8 | 4 | 168.517 / 152.910 | 168.382 / 154.927 |
| 1024 | 5 | uint8 | 8 | 340.973 / 291.825 | 342.739 / 296.125 |
| 1024 | 5 | uint8 | 16 | 688.525 / 576.542 | 687.726 / 576.127 |
| 1024 | 5 | float32 | 2 | 65.645 / 60.147 | 68.239 / 61.714 |
| 1024 | 5 | float32 | 4 | 138.909 / 104.803 | 140.420 / 107.208 |
| 1024 | 5 | float32 | 8 | 244.168 / 193.198 | 245.141 / 193.487 |
| 1024 | 5 | float32 | 16 | 480.284 / 373.323 | 482.111 / 372.077 |

</details>

## Review-driven guards and permanent benchmark coverage

- Singleton batches now bypass batch setup after public Compose measurements showed 8–45% regressions for representative 256–1024 inputs.
- Zero-effect batches bypass pattern expansion and arithmetic while retaining copy/clipping semantics; float32 uses a one-pass clipped copy.
- For uint8 `1024x1024x5` brightness-only batches, the inherited fallback measured 41.4 ms versus 65.8 ms at N=2 and 80.9 ms versus 118.6 ms at N=4, so that workload is guarded back to per-image execution. N=3 and N=5 benchmark cases cover both sides of the boundary.
- Permanent ASV coverage contains 252 direct, 252 Compose-images, and 40 Compose-volume timing cases (544 total), plus 12 peak-memory cases. It includes normal, singleton, zero-effect, brightness-only, threshold-boundary, and large-volume sentinels.
- The changed-pixel performance profile explicitly selects all three Plasma batch timing classes, with a regression test for the resolved regex.

## Validation

- Focused functional parity/non-mutation test: 48 passed.
- Pyrefly under the CI Python 3.10 dependency profile: 0 errors.
- Current head `eeb9973`: hosted CI completed with 35 successful, 12 intentionally skipped, and 0 failing or pending checks, including correctness, compatibility, coverage, PyTorch, type checks, benchmark evidence, and core performance smoke.

## Summary by Sourcery

Optimize PlasmaBrightnessContrast batch processing while maintaining exact per-image behavior and public API compatibility.

Bug Fixes:
- Preserve per-image contrast reduction semantics for PlasmaBrightnessContrast batches instead of incorrectly using a batch-wide mean.

Enhancements:
- Optimize PlasmaBrightnessContrast batch processing with adaptive single-image, dedicated, shared-map loop, and specialized fallback routes while preserving output ownership, layout, dtype, and clipping behavior.

Tests:
- Add exact parity, non-mutation, output-contract, edge-case, and dispatcher-route coverage for PlasmaBrightnessContrast image and volume batches.
- Add comprehensive timing and peak-memory benchmark coverage for PlasmaBrightnessContrast batch paths and verify changed-pixel benchmark selection.

### CODE REVIEW

File: albumentations/augmentations/pixel/_functional_illumination.py [hunk 1]
```diff
@@ -172,6 +180,219 @@ def apply_plasma_brightness_contrast(
...
+    """Process a small float32 batch as one array while retaining one global mean per image, reducing Python
+    dispatch where whole-batch memory traffic remains bounded.
+    """
+    work = np.ascontiguousarray(images)
+    image_pattern = _plasma_pattern_for_batch(images, plasma_pattern)
+    brightness_map = None
+
+    if brightness_factor != 0:
+        brightness_map = np.multiply(image_pattern, np.float32(brightness_factor))
+        work = cast("ImageFloat32", add(work, brightness_map, inplace=True))
```
[the remark this answers is not in the document]
Comment (ayasyrev) [reply albumentations/augmentations/pixel/_functional_illumination.py]: Thanks for flagging this. I traced the concrete Albucore dispatch and verified that this path does not mutate the caller input. Every operand here is a multidimensional array, so `add(..., inplace=True)` and `multiply(..., inplace=True)` route to `add_array` / `multiply_by_array`; those route to `add_numpy` / `multiply_numpy`, whose NumPy ufunc call has no `out=` argument and therefore returns an owned array. The later in-place clip operates on that owned result. This covers brightness-only, contrast-only, and combined execution in both the dedicated and loop kernels. The focused parity/ownership test passes all 48 cases and explicitly asserts exact input preservation plus no output memory sharing; an additional dedicated/loop probe also found neither mutation nor sharing. Adding an eager copy would introduce a redundant full-array pass on this hot path, so I am keeping the current allocation boundary.

**Comment by ternaus** [comment]
@codex
